File: packages/alpaca-option/src/display.ts
import { display as timeDisplay } from '@alpaca/time';

import { canonicalContract, type ContractInput } from './contract';
import type { ContractDisplay, OptionRight, OptionRightCode } from './types';

export function formatStrike(strike: number): string {
  let text = strike.toFixed(3);
  while (text.includes('.') && text.endsWith('0')) {
    text = text.slice(0, -1);
  }
  if (text.endsWith('.')) {
    text = text.slice(0, -1);
  }
  return text;
}

export function positionStrike(position?: {
  contract?: ContractInput;
} | null): string {
  const contract = canonicalContract(position?.contract);
  return contract ? formatStrike(contract.strike) : '-';
}

export function compactContract(
  contract?: ContractInput,
  expirationFormat: 'mm-dd' | 'yy-mm-dd' | 'yymmdd' = 'mm-dd',
): string {
  return contractDisplay(contract, expirationFormat)?.compact ?? '-';
}

export function contractDisplay(
  contractInput?: ContractInput,
  expirationFormat: 'mm-dd' | 'yy-mm-dd' | 'yymmdd' = 'mm-dd',
): ContractDisplay | null {
  const contract = canonicalContract(contractInput);
  if (!contract) {
    return null;
  }

  const strike = formatStrike(contract.strike);
  const optionRight = optionRightCode(contract.option_right);
  const expiration = timeDisplay.compact(contract.expiration_date, expirationFormat);
  return {
    strike,
    expiration,
    compact: `${strike}${optionRight}@${expiration}`,
    optionRightCode: optionRight,
  };
}

export function optionRightCode(optionRight: OptionRight): OptionRightCode {
  return optionRight === 'call' ? 'C' : 'P';
}

File: crates/alpaca-option/src/display.rs
use alpaca_time::display;

use crate::types::{ContractDisplay, OptionContract, OptionPosition, OptionRight, OptionRightCode};

pub fn format_strike(strike: f64) -> String {
    let mut text = format!("{strike:.3}");
    while text.contains('.') && text.ends_with('0') {
        text.pop();
    }
    if text.ends_with('.') {
        text.pop();
    }
    text
}

pub fn position_strike(position: &OptionPosition) -> String {
    format_strike(position.contract_info().strike)
}

pub fn compact_contract(contract: &OptionContract, expiration_style: Option<&str>) -> String {
    contract_display(contract, expiration_style).compact
}

pub fn contract_display(contract: &OptionContract, expiration_style: Option<&str>) -> ContractDisplay {
    let strike = format_strike(contract.strike);
    let option_right_code = contract.option_right.code_string();
    let expiration = display::compact(&contract.expiration_date, expiration_style.unwrap_or("mm-dd"));

    ContractDisplay {
        strike: strike.clone(),
        expiration: expiration.clone(),
        compact: format!("{strike}{}@{expiration}", contract.option_right.code()),
        option_right_code,
    }
}

pub fn option_right_code(option_right: &OptionRight) -> OptionRightCode {
    option_right.code_string()
}
